flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -1937,6 +1966,14 @@ mlir::Value IntrinsicLibrary::genDim(mlir::Type resultType,
   return builder.create<mlir::SelectOp>(loc, cmp, diff, zero);
 }
 
+// DOT_PRODUCT
+fir::ExtendedValue
+IntrinsicLibrary::genDotProduct(mlir::Type resultType,
+                                llvm::ArrayRef<fir::ExtendedValue> args) {
+  return genDotProd(Fortran::lower::genDotProduct, resultType, builder, loc,
```

Comment:
Not sure if you need the templated function, genDotProd(). Are there other intrinsics that take two rank one arrays as arguments?  Matmul comes close. However, matmul can take rank one or rank two array arguments. Dot product arguments are just rank one. Whoever implements matmul can possibly add it to genDotProd(). They probably should rename the function if they do it. @jeanPerier @schweitzpgi  what are your thoughts?

Reply:
It's not necessary to create a template function that will only be used once, but it won't really hurt anything on the other hand. As you say, this might be leveraged by matmul with some changes eventually as well.

Reply:
And since we can rely on the rank checking in the front-end, it probably will be pretty easy to add matmul to the templated function too. So, I think it's fine to leave it as a templated function.